pandas/indexes/base.py

```diff
@@ -1403,46 +1407,31 @@ def _ensure_compat_append(self, other):
             to_concat.append(other)
 
         for obj in to_concat:
-            if (isinstance(obj, Index) and obj.name != name and
-                    obj.name is not None):
-                name = None
-                break
+            if not isinstance(obj, Index):
+                raise TypeError('all inputs must be Index')
```

Comment:
Yes, I agree docstring intends first one. I'm adding an validation to check all elements are ``Index`` if input is list-like.

I think it is regarded as a bug fix, because it is similar to ``CategiricalIndex.append`` accepts (flat) list of labels.